### PR TITLE
[14.0][IMP+FIX] base_location

### PR DESCRIPTION
**Pull request by ivantodorovich**
A few usability improvements on `base_location`

Also fixes a bug introduced here: https://github.com/OCA/partner-contact/commit/5fd93fb8d138cb9a650410d49745b0777be43be0, more details on the commit messages

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Please indicate the steps to reproduce the problem before the PR.

**Comment by ivantodorovich** [comment]
> Please indicate the steps to reproduce the problem before the PR.

Ohh yes, sorry :sweat_smile: 

1. Configure a country with `enforce_cities`
2. Create a new partner, and choose the country you've configured

Expected: City should be a one2many (city_id) and not a char field (city).
Result: Not that


image

image



As explained on the commit message, this is due to the fact that we're adding city_id onto the view xml, and `_fields_view_get_address` in `base_address_city` won't do its magic if that's the case:

https://github.com/odoo/odoo/blob/e06aaecf9e8a590604818b09d23fe1de410e6762/addons/base_address_city/models/res_partner.py#L27-L33

File: base_location/models/res_company.py [hunk 1]
```diff
@@ -48,24 +47,22 @@ def _get_company_address_field_names(self):
...
+        self = self.with_context(skip_check_zip=True)
         for company in self:
-            company.with_context(
-                skip_check_zip=True
-            ).partner_id.city_id = company.city_id
+            company.partner_id.city_id = company.city_id
 
     def _inverse_zip_id(self):
+        self = self.with_context(skip_check_zip=True)
         for company in self:
```
Comment (pedrobaeza) [inline base_location/models/res_company.py]:
```suggestion
        for company in self.with_context(skip_check_zip=True):
```

File: base_location/models/res_company.py [hunk 2]
```diff
@@ -48,24 +47,22 @@ def _get_company_address_field_names(self):
         return res
 
     def _inverse_city_id(self):
+        self = self.with_context(skip_check_zip=True)
         for company in self:
```
Comment (pedrobaeza) [inline base_location/models/res_company.py]: Save one line and avoid the overwriting of the variable:
```suggestion
        for company in self.with_context(skip_check_zip=True):
```

File: base_location/models/res_company.py [hunk 3]
```diff
@@ -48,24 +47,22 @@ def _get_company_address_field_names(self):
...
-            company.with_context(skip_check_zip=True).partner_id.zip_id = company.zip_id
+            company.partner_id.zip_id = company.zip_id
 
     def _inverse_state(self):
-        return super(
-            ResCompany, self.with_context(skip_check_zip=True)
-        )._inverse_state()
+        self = self.with_context(skip_check_zip=True)
```
Comment (pedrobaeza) [inline base_location/models/res_company.py:58]: This is not correct, as it's a local variable and it's not pass upstream. Please revert.

File: base_location/models/res_company.py [hunk 4]
```diff
@@ -48,24 +47,22 @@ def _get_company_address_field_names(self):
...
+        return super()._inverse_state()
 
     def _inverse_country(self):
-        return super(
-            ResCompany, self.with_context(skip_check_zip=True)
-        )._inverse_country()
+        self = self.with_context(skip_check_zip=True)
```
Comment (pedrobaeza) [inline base_location/models/res_company.py:62]: The same.

File: base_location/models/res_company.py (hunk 3 above)
Comment (ivantodorovich) [reply base_location/models/res_company.py:58]: It is passed :)

Comment (pedrobaeza) [reply base_location/models/res_company.py:58]: Well, I prefer not to do such things that depends on the Python implementation. Python zen: explicit is better than implicit.

Comment (ivantodorovich) [reply base_location/models/res_company.py:58]: IMO it's explicit but ok, it's changed

**Review by pedrobaeza — APPROVED** [review]

**Review by simahawk — APPROVED** [review]
LG

**Pull request merged**